### PR TITLE
Add local file path support to file urls

### PR DESCRIPTION
**Pull request by ArtemShatokhin**
> [!NOTE]
> Enables attaching local files via `file_urls` when `allowed_local_file_dirs` is configured, with strict path allowlisting and improved error handling across sync, streaming, and AG-UI endpoints.
> 
> - Adds `allowed_local_file_dirs` to `run_fastapi` and passes normalized allowlist to `get_response`, `get_response_stream`, and AG-UI handlers
> - Refactors file handling into `fastapi_utils/file_handler.py`: supports http(s), file:// URIs, absolute/UNC paths; validates extensions; enforces allowlist; uploads to OpenAI and waits for processing
> - Extends request models to include `file_ids`/`file_urls` for AG-UI (`RunAgentInputCustom`) and base requests
> - Improves AG-UI handler to emit RUN_STARTED/ERROR/FINISHED on `file_urls` failures; streaming endpoint now returns SSE error for bad downloads; both include `file_ids_map`
> - Updates documentation: parameter tables, endpoint reference, payload schema, file attachments (including local paths), cancellation behavior, and auth notes
> - Adds comprehensive tests: unit tests for file handler edge cases and AG-UI error path; integration tests for remote and local attachments, disallowed local access, streaming errors
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0fb4394ecdfa99a8c10c1522232583350225abc4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by nicko-ai** [comment]
@codex review

**Comment by nicko-ai** [comment]
@codex review

File: docs/additional-features/fastapi-integration.mdx
Comment (nicko-ai) [inline docs/additional-features/fastapi-integration.mdx:1]: introduce subsections:
- endpoint (run_fastapi) parameters
- input parameters (payload fields)
- authentication : bearer token
- user_context - add to the payload description (accordion)
- response format
- Implementation reference

Do not repeat the Endpoint Structure 3 times (reduce to 1-2)

Remove "If enable_logging=True, a /get_logs endpoint is also added."

File: docs/additional-features/fastapi-integration.mdx [hunk 1]
```diff
@@ -33,46 +33,141 @@ agency = Agency(agent, name="test_agency")
...
+</Accordion>
+
+<Accordion title="Endpoint reference" defaultOpen={false}>
+
+- Agencies are served at:
+  - `/your_agency_name/get_response` (POST)
+  - `/your_agency_name/get_response_stream` (POST, streaming responses)
+  - `/your_agency_name/cancel_response_stream` (POST; not registered when `enable_agui=True`)
+  - `/your_agency_name/get_metadata` (GET)
+  - `/your_agency_name/get_logs` (GET; when `enable_logging=True`)
```
Comment (nicko-ai) [inline docs/additional-features/fastapi-integration.mdx]: incorrect path in the docs, please fix here and in other place where applicable:

```suggestion
  - `/get_logs` (GET; when `enable_logging=True`)
```

**Review by bonk1t — CHANGES_REQUESTED** [review]
Key issues and AGENTS.md violations:

- All file upload errors are mapped to HTTP 404. This turns bad input and permission errors into “not found,” which is likely the wrong status and makes client debugging harder. src/agency_swarm/integrations/fastapi_utils/endpoint_handlers.py:121 and src/agency_swarm/integrations/fastapi_utils/endpoint_handlers.py:193
  Suggested solution: Add a small exception-to-status mapping (PermissionError → 403, FileNotFoundError → 404, IsADirectoryError/ValueError → 400, httpx.HTTPStatusError → 404, default → 500) and use it in both endpoints so input and permission errors are not returned as 404. Update the local-path denial test to assert 403 while keeping the missing remote file case at 404.

- Allowlist validation is inconsistent: when allowed_local_dirs is a list of Path, existence/dir checks are skipped, but string paths are validated. This can silently allow bad allowlists. src/agency_swarm/integrations/fastapi_utils/file_handler.py:216 and src/agency_swarm/integrations/fastapi_utils/file_handler.py:157
  Suggested solution: Remove the Path-only branch and always run _normalize_allowed_dirs so both Path and str entries go through the same exists/is_dir checks. Add a unit test that passes a Path to a missing directory and asserts FileNotFoundError to lock the behavior.

- New integration test uses a fixed port and in-function imports. The fixed port (8081) can collide in CI, and imports inside the fixture violate AGENTS.md. tests/integration/fastapi/test_fastapi_file_processing.py:159 and tests/integration/fastapi/test_fastapi_file_processing.py:161
  Suggested solution: Use _get_free_tcp_port() for the no-local fixture, pass that port into run_fastapi/uvicorn, and build the health-check URL from it. Move threading and run_fastapi imports to the module top so imports are top-level.

- AGENTS.md compliance issues in new helpers: new functions lack type hints and private helpers are placed before public upload_from_urls, which breaks the repo’s ordering rule. src/agency_swarm/integrations/fastapi_utils/file_handler.py:107 and src/agency_swarm/integrations/fastapi_utils/file_handler.py:196
  Suggested solution: Add type hints to any helper functions in file_handler.py that were added without them, and move upload_from_urls above the private helpers (or move the helpers below it). This restores public-first ordering and satisfies the type-hint rule.

File: docs/additional-features/fastapi-integration.mdx [hunk 2]
```diff
@@ -237,94 +312,85 @@ response = requests.post(cancel_url, json=payload)
...
-See `examples/fastapi_integration/server.py` for a complete multi-agent + tool setup.
+To support passing local filepaths in the `file_urls` field, set `allowed_local_file_dirs` on `run_fastapi`:
 
----
+```python
+run_fastapi(agencies=..., allowed_local_file_dirs=["/data/uploads"])
+```
 
-## File Attachments
+Then pass file paths in `file_urls` (supports both local and absolute paths):
```
Comment (nicko-ai) [inline docs/additional-features/fastapi-integration.mdx]: "both local and absolute paths" - what does it mean?

**Comment by ArtemShatokhin** [comment]
@bonk1t had to revert the updated error handling due to it causing issues on SaaS

**Comment by nicko-ai** [comment]
@codex review

File: src/agency_swarm/integrations/fastapi_utils/file_handler.py [hunk 3]
```diff
@@ -13,113 +14,233 @@
...
+            _validate_local_file(path, path_or_url)
+            local_files[name] = path
+            continue
+
+        # Absolute filesystem paths
+        path = Path(path_or_url)
+        if path.is_absolute():
+            _ensure_path_allowed(path, _get_allowed_dirs())
+            _validate_local_file(path, path_or_url)
+            local_files[name] = path
```
[the remark this answers is not in the document]
Comment (nicko-ai) [reply src/agency_swarm/integrations/fastapi_utils/file_handler.py:88]: @ArtemShatokhin The docs say local file_urls support both relative and absolute paths (docs/additional-features/fastapi-integration.mdx:388), but upload_from_urls only treats file://, Windows drive, or absolute paths as local; relative paths fall into remote URL validation and raise “Unsupported URL scheme: none” (src/agency_swarm/integrations/fastapi_utils/file_handler.py:74 and src/agency_swarm/integrations/fastapi_utils/file_handler.py:85). This looks like a real mismatch. Do you want to support relative paths (resolved against allowed_local_file_dirs) or should we tighten the docs and model descriptions to say absolute paths only?

[the remark this answers is not in the document]
Comment (nicko-ai) [reply src/agency_swarm/integrations/fastapi_utils/file_handler.py:88]: One more doc mismatch: the FastAPI docs still say invalid file_urls return HTTP 404, but the code now returns HTTP 200 with an error body for both non‑streaming and streaming endpoints (src/agency_swarm/integrations/fastapi_utils/endpoint_handlers.py:118 and src/agency_swarm/integrations/fastapi_utils/endpoint_handlers.py:193). Also, request_models still documents “absolute local paths only” (src/agency_swarm/integrations/fastapi_utils/request_models.py:30), which conflicts with the docs claim about relative paths. Should we update the docs/model text to match the current behavior?

[the remark this answers is not in the document]
Comment (ArtemShatokhin) [reply src/agency_swarm/integrations/fastapi_utils/file_handler.py:88]: @bonk1t No, relative paths are not supported, it was a typo in docs.

**Comment by nicko-ai** [comment]
@codex review
